### PR TITLE
feat: B/P Detailed Errors, race condition fixes, and preliminary support for the Customer Account API

### PR DESCRIPTION
**Pull request by nickevansuk**
This PR includes:

1. Very early work to add preliminary `CustomerAccount` support based on the ongoing conversations around the [Customer Accounts proposal](https://github.com/openactive/customer-accounts), to inform the OpenActive specification work for Customer Accounts with implementation feedback.
2. B/P detailed errors (https://github.com/openactive/open-booking-api/issues/218)
3. Prevention of race conditions from #164

---

Migration required for users of the library as follows:
- Find-and-replace (case sensitive) all instances of `SellerIdComponents` with `SimpleIdComponents` 
- Find-and-replace all instances of `SellerIdLong` with `IdLong`
- Find-and-replace all instances of `SellerIdGuid` with `IdGuid`
- Find-and-replace all instances of `SellerIdString` with `IdString`
- Find-and-replace all instances of `(string clientId, Uri sellerId) = User.GetAccessTokenOpenBookingClaims();` with `(string clientId, Uri sellerId, _) = User.GetAccessTokenOpenBookingClaims();`

### CODE REVIEW

File: Examples/BookingSystem.AspNetCore/Properties/launchSettings.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
       "launchUrl": "https://localhost:5001/openactive",
       "applicationUrl": "https://localhost:5001",
       "environmentVariables": {
-        "ASPNETCORE_ENVIRONMENT": "Development",
+        "ASPNETCORE_ENVIRONMENT": "no-auth",
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetCore/Properties/launchSettings.json:13]: Suspect this shouldn't have been left in?

File: Examples/BookingSystem.AspNetCore/Stores/OrderStore.cs [hunk 2]
```diff
@@ -449,21 +449,33 @@ public override async Task<Order> GetOrderStatus(OrderIdComponents orderId, Simp
...
+                    OrderedItem = orderItem.OpportunityJsonLdId,
+                    OrderItemStatus =
+                                orderItem.Status == BookingStatus.Confirmed ? OrderItemStatus.OrderItemConfirmed :
+                                orderItem.Status == BookingStatus.CustomerCancelled ? OrderItemStatus.CustomerCancelled :
+                                orderItem.Status == BookingStatus.SellerCancelled ? OrderItemStatus.SellerCancelled :
+                                orderItem.Status == BookingStatus.Attended ? OrderItemStatus.AttendeeAttended :
+                                orderItem.Status == BookingStatus.Absent ? OrderItemStatus.AttendeeAbsent : (OrderItemStatus?)null,
+                    Attendee = hasAttendeeDetails ? new Person
+                    {
+                        GivenName = orderItem.AttendeeGivenName ?? null,
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetCore/Stores/OrderStore.cs:473]: Not sure `?? null` has any affect here? If it's already null, then null will be assigned

**Review by nickevansuk — COMMENTED** [review]
A few notes

File: Examples/BookingSystem.AspNetCore/Properties/launchSettings.json (hunk 1 above)
Comment (civsiv) [reply Examples/BookingSystem.AspNetCore/Properties/launchSettings.json:13]: Have removed that in the next commit

File: Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs [hunk 3]
```diff
@@ -838,7 +838,11 @@ public class BookedOrderItemInfo
             Uri opportunityJsonLdId,
             Uri offerJsonLdId,
             long numberOfSpaces,
-            bool proposal
+            bool proposal,
+            List<string> attendeeEmail,
```
[the remark this answers is not in the document]
Comment (civsiv) [reply Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:842]: Was following the pattern already set for speed

File: Examples/BookingSystem.AspNetFramework/Web.config [hunk 4]
```diff
@@ -1,380 +1,398 @@
...
+        <add assembly="System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" />
+        <add assembly="System.Data.DataSetExtensions, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" />
+        <add assembly="System.Xml.Linq, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" />
+        <add assembly="System.Web, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
+        <add assembly="System.Web.Abstractions, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
+        <add assembly="System.Web.Routing, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
+        <add assembly="System.Xml, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" />
+        <add assembly="System.Configuration, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
+        <add assembly="System.Runtime.Serialization, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" />
+        <add assembly="System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetFramework/Web.config:50]: Revert dependency changes

```suggestion
```

File: Examples/BookingSystem.AspNetCore/Stores/FacilityStore.cs [hunk 5]
```diff
@@ -286,7 +286,7 @@ protected override async Task TriggerTestAction(OpenBookingSimulateAction simula
         // Similar to the RPDE logic, this needs to render and return an new hypothetical OrderItem from the database based on the supplied opportunity IDs
         protected override async Task GetOrderItems(List<OrderItemContext<FacilityOpportunity>> orderItemContexts, StoreBookingFlowContext flowContext, OrderStateContext stateContext)
         {
-            // Note the implementation of this method must also check that this OrderItem is from the Seller specified by context.SellerIdComponents (this is not required if using a Single Seller)
+            // Note the implementation of this method must also check that this OrderItem is from the Seller specified by context.SimpleIdComponents (this is not required if using a Single Seller)
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetCore/Stores/FacilityStore.cs:289]:
```suggestion
            // Note the implementation of this method must also check that this OrderItem is from the Seller specified by context.SellerId (this is not required if using a Single Seller)
```

File: Examples/BookingSystem.AspNetCore/Stores/FacilityStore.cs [hunk 6]
```diff
@@ -555,13 +569,27 @@ protected override async ValueTask ProposeOrderItems(List<OrderItemContext<Facil
...
+                    ctxGroup
+                        .Where(x => x.RequestOrderItem.Attendee != null)
+                        .Select(x => new FakeDbPerson
+                        {
+                            GivenName = x.RequestOrderItem.Attendee.GivenName,
+                            FamilyName = x.RequestOrderItem.Attendee.FamilyName,
+                            Email = x.RequestOrderItem.Attendee.Email,
+                            Telephone = x.RequestOrderItem.Attendee.Telephone,
+                        })
+                        .ToList(),
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetCore/Stores/FacilityStore.cs:588]: @civsiv will this not cause issues with the attendees not being stored against the correct OrderItems if only some attendees are specified.

e.g. at P

1: Bob
2: Null
3: Alex

then B would be

1: Bob
2: Alex
3: Null

So we'd actually want something more like the below:?

```suggestion
                        .Select(x => x.RequestOrderItem.Attendee == null ? null : new FakeDbPerson
                        {
                            GivenName = x.RequestOrderItem.Attendee.GivenName,
                            FamilyName = x.RequestOrderItem.Attendee.FamilyName,
                            Email = x.RequestOrderItem.Attendee.Email,
                            Telephone = x.RequestOrderItem.Attendee.Telephone,
                        })
                        .ToList(),
```

File: Examples/BookingSystem.AspNetCore/Stores/SessionStore.cs [hunk 7]
```diff
@@ -556,13 +556,27 @@ protected override async ValueTask BookOrderItems(List<OrderItemContext<SessionO
...
+                    ctxGroup
+                        .Where(x => x.RequestOrderItem.Attendee != null)
+                        .Select(x => new FakeDbPerson
+                        {
+                            GivenName = x.RequestOrderItem.Attendee.GivenName,
+                            FamilyName = x.RequestOrderItem.Attendee.FamilyName,
+                            Email = x.RequestOrderItem.Attendee.Email,
+                            Telephone = x.RequestOrderItem.Attendee.Telephone,
+                        })
+                        .ToList(),
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetCore/Stores/SessionStore.cs:575]: Same issue as above?

File: Examples/BookingSystem.AspNetCore/Stores/SessionStore.cs [hunk 8]
```diff
@@ -607,13 +621,27 @@ protected override async ValueTask ProposeOrderItems(List<OrderItemContext<Sessi
...
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetCore/Stores/SessionStore.cs:640]: Same issue as above?

File: Examples/BookingSystem.AspNetCore/Stores/FacilityStore.cs [hunk 9]
```diff
@@ -504,13 +504,27 @@ protected override async ValueTask BookOrderItems(List<OrderItemContext<Facility
...
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetCore/Stores/FacilityStore.cs:523]: Same issue as below?

File: Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs [hunk 10]
```diff
@@ -882,8 +892,14 @@ bool proposal
                     BarCodeText = thisClass.AttendanceMode != AttendanceMode.Online ? Faker.Random.String(length: 10, minChar: '0', maxChar: '9') : null,
                     MeetingUrl = thisClass.AttendanceMode != AttendanceMode.Offline ? new Uri(Faker.Internet.Url()) : null,
                     MeetingId = thisClass.AttendanceMode != AttendanceMode.Offline ? Faker.Random.String(length: 10, minChar: '0', maxChar: '9') : null,
-                    MeetingPassword = thisClass.AttendanceMode != AttendanceMode.Offline ? Faker.Random.String(length: 10, minChar: '0', maxChar: '9') : null
+                    MeetingPassword = thisClass.AttendanceMode != AttendanceMode.Offline ? Faker.Random.String(length: 10, minChar: '0', maxChar: '9') : null,
+                    AttendeeEmail = attendees.Count > 0 ? attendees[i].Email : null,
```
Comment (nickevansuk) [inline Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:896]: Does this check make sense? Could we be out-of-bounds of the array if `i` was 2 and the attendee count was 1 (due to the `Where` clause flagged earlier).

Perhaps something more defensive such as:

```suggestion
                    AttendeeEmail = attendees.Count > i ? attendees[i].Email : null,
``` 

Also note any amends to this pattern need to be replicated below also

**Review by nickevansuk — COMMENTED** [review]
@civsiv a few more minor fixes here

File: Examples/BookingSystem.AspNetFramework/Stores/SessionStore.cs [hunk 11]
```diff
@@ -306,7 +306,7 @@ protected override async Task TriggerTestAction(OpenBookingSimulateAction simula
         // Similar to the RPDE logic, this needs to render and return an new hypothetical OrderItem from the database based on the supplied opportunity IDs
         protected override async Task GetOrderItems(List<OrderItemContext<SessionOpportunity>> orderItemContexts, StoreBookingFlowContext flowContext, OrderStateContext stateContext)
         {
-            // Note the implementation of this method must also check that this OrderItem is from the Seller specified by context.SellerIdComponents (this is not required if using a Single Seller)
+            // Note the implementation of this method must also check that this OrderItem is from the Seller specified by context.SimpleIdComponents (this is not required if using a Single Seller)
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetFramework/Stores/SessionStore.cs:309]:
```suggestion
            // Note the implementation of this method must also check that this OrderItem is from the Seller specified by context.SellerId (this is not required if using a Single Seller)
```

File: Examples/BookingSystem.AspNetCore/Stores/SessionStore.cs [hunk 12]
```diff
@@ -306,7 +306,7 @@ protected override async Task TriggerTestAction(OpenBookingSimulateAction simula
         // Similar to the RPDE logic, this needs to render and return an new hypothetical OrderItem from the database based on the supplied opportunity IDs
         protected override async Task GetOrderItems(List<OrderItemContext<SessionOpportunity>> orderItemContexts, StoreBookingFlowContext flowContext, OrderStateContext stateContext)
         {
-            // Note the implementation of this method must also check that this OrderItem is from the Seller specified by context.SellerIdComponents (this is not required if using a Single Seller)
+            // Note the implementation of this method must also check that this OrderItem is from the Seller specified by context.SimpleIdComponents (this is not required if using a Single Seller)
```
Comment (nickevansuk) [inline Examples/BookingSystem.AspNetCore/Stores/SessionStore.cs:309]:
```suggestion
            // Note the implementation of this method must also check that this OrderItem is from the Seller specified by context.SellerId (this is not required if using a Single Seller)
```

File: Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs (hunk 10 above)
Comment (civsiv) [reply Fakes/OpenActive.FakeDatabase.NET/FakeBookingSystem.cs:896]: I'm going to change this to be a serialised string which will remove all of these issues

**Pull request merged**